src/replication/adapter/jfrog/adapter.go

```diff
@@ -0,0 +1,312 @@
+	info = &model.RegistryInfo{
+		Type: model.RegistryTypeJfrogArtifactory,
+		SupportedResourceTypes: []model.ResourceType{
+			model.ResourceTypeImage,
+		},
+		SupportedResourceFilters: []*model.FilterStyle{
+			{
+				Type:  model.FilterTypeName,
+				Style: model.FilterStyleTypeText,
+			},
+			{
+				Type:  model.FilterTypeTag,
+				Style: model.FilterStyleTypeText,
+			},
+		},
+		SupportedTriggers: []model.TriggerType{
+			model.TriggerTypeManual,
+			model.TriggerTypeScheduled,
+		},
+	}
+	return
+}
+
+// AdapterFactory is the factory for jfrog artifactory adapter
+func AdapterFactory(registry *model.Registry) (adp.Adapter, error) {
+	dockerRegistryAdapter, err := native.NewAdapter(registry)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		modifiers = []modifier.Modifier{
+			&auth.UserAgentModifier{
+				UserAgent: adp.UserAgentReplication,
+			}}
+		authorizer modifier.Modifier
+	)
+	if registry.Credential != nil {
+		authorizer = auth.NewBasicAuthCredential(
+			registry.Credential.AccessKey,
+			registry.Credential.AccessSecret)
+		modifiers = append(modifiers, authorizer)
+	}
+
+	return &adapter{
+		Adapter:  dockerRegistryAdapter,
+		registry: registry,
+		client: common_http.NewClient(
+			&http.Client{
+				Transport: util.GetHTTPTransport(registry.Insecure),
+			},
+			modifiers...,
+		),
+	}, nil
+
+}
+
+// PrepareForPush creates local docker repository in jfrog artifactory
+func (a *adapter) PrepareForPush(resources []*model.Resource) error {
+	var namespaces []string
+	for _, resource := range resources {
+		if resource == nil {
+			return errors.New("the resource cannot be null")
+		}
+		if resource.Metadata == nil {
+			return errors.New("the metadata of resource cannot be null")
+		}
+		if resource.Metadata.Repository == nil {
+			return errors.New("the namespace of resource cannot be null")
+		}
+		if len(resource.Metadata.Repository.Name) == 0 {
+			return errors.New("the name of namespace cannot be null")
+		}
+		path := strings.Split(resource.Metadata.Repository.Name, "/")
+		if len(path) > 0 {
+			namespaces = append(namespaces, path[0])
+		}
+	}
+
+	repositories, err := a.getLocalRepositories()
+	if err != nil {
+		log.Errorf("Get local repositories error: %v", err)
+		return err
+	}
+
+	existedRepositories := make(map[string]struct{})
+	for _, repo := range repositories {
+		existedRepositories[repo.Key] = struct{}{}
+	}
+
+	for _, namespace := range namespaces {
+		if _, ok := existedRepositories[namespace]; ok {
+			log.Debugf("Namespace %s already existed in remote, skip create it", namespace)
+		} else {
+			err = a.createNamespace(namespace)
+			if err != nil {
+				log.Errorf("Create Namespace %s error: %v", namespace, err)
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (a *adapter) getLocalRepositories() ([]*repository, error) {
+	var repositories []*repository
+	url := fmt.Sprintf("%s/artifactory/api/repositories?type=local&packageType=docker", a.registry.URL)
```

Comment:
It will get all repositories, without pagination ?

Reply:
I think it will get all，from offical api document, this interface just provide two query parameters(type and packageType), without pagination definition.

Reply:
https://www.jfrog.com/confluence/display/RTF/Artifactory+REST+API#ArtifactoryRESTAPI-GetRepositories